src/calculate_bucket.hpp

```diff
@@ -77,7 +77,9 @@ class F1Calculator {
     inline F1Calculator(uint8_t k, const uint8_t* orig_key)
     {
         uint8_t enc_key[32];
+        size_t buf_blocks = cdiv((k + kExtraBits) << kBatchSizes, kF1BlockSizeBits) + 2;
         this->k_ = k;
+        this->buf_ = new uint8_t[buf_blocks * kF1BlockSizeBits / 8];
```

Comment:
`buf_` is only used in `CalculateBuckets()`, and it's overwritten every time that function is called. It's not obvious why it needs to be allocated in the constructor, why not allocate it temporarily in `CalculateBuckets()`?

Additionally, the size of this allocation is computed in a different way than the `n_blks` is computed. I imagine they end up being the same, since the test pass. It's not obvious that they are by reading the code though.

Reply:
The reason for allocating it here rather than in `CalculateBuckets` is that the latter function is "hot" and must run as fast as possible (it is the one being optimized in this commit).

Regarding the size of the allocation, there's a comment to `CalculateBuckets`, which states that `n` must not exceed `1 << kBatchSizes`. The allocation size is chosen so that it's sufficient for any `n` within the specified range.

Reply:
I see. so this is sort of like a "scratch buffer", that's reused to avoid the cost of heap allocation